src/prefect/task_engine.py

```diff
@@ -690,8 +692,11 @@ async def wait_until_ready(self):
         if scheduled_time := self.state.state_details.scheduled_time:
             sleep_time = (scheduled_time - pendulum.now("utc")).total_seconds()
             await anyio.sleep(sleep_time if sleep_time > 0 else 0)
+            new_state = Retrying() if self.state.name == "AwaitingRetry" else Running()
+            if self.state.name == "AwaitingRetry":
+                self.task_run.run_count += 1
             self.set_state(
-                Retrying() if self.state.name == "AwaitingRetry" else Running(),
+                new_state,
```

Comment:
Jw: would it be simpler to have an if new_state.is_running() then increment run count statement in set_state or will that have edge cases?

Reply:
Yup, that does appear to work. Updated in https://github.com/PrefectHQ/prefect/pull/15436/commits/3fb2d68c112a8ef298819ccbecff858990443fe6